controlpanel/api/models/tool.py

```diff
@@ -0,0 +1,131 @@
+import logging
+import secrets
+
+from django.conf import settings
+from django.contrib.postgres.fields import JSONField
+import django.core.exceptions
+from django.db import models
+from django.db.models import Q
+from django_extensions.db.models import TimeStampedModel
+
+from controlpanel.api import cluster
+
+
+log = logging.getLogger(__name__)
+
+
+class Tool(TimeStampedModel):
+    """
+    Represents a software tool available to users of the platform. An instance
+    of Tool is an item in the Software Catalogue - not a user's deployed
+    instance of a tool.
+    """
+
+    description = models.TextField(blank=True)
+    chart_name = models.CharField(max_length=100, blank=False)
+    name = models.CharField(max_length=100, blank=False)
+    values = JSONField(default=dict)
+    version = models.CharField(max_length=100, blank=False)
+
+    class Meta(TimeStampedModel.Meta):
+        db_table = "control_panel_api_tool"
+        ordering = ('name',)
+
+    def __repr__(self):
+        return f'<Tool: {self.chart_name} {self.version}>'
+
+    def url(self, user):
+        return f'https://{user.slug}-{self.chart_name}.{settings.TOOLS_DOMAIN}/'
+
+
+class ToolDeploymentManager:
+    """
+    Emulates a Django model manager
+    """
+
+    def create(self, *args, **kwargs):
+        tool_deployment = ToolDeployment(*args, **kwargs)
+        tool_deployment.save()
+        return tool_deployment
+
+    def filter(self, **kwargs):
+        user = kwargs.get('user')
+        filter = Q(chart_name=None)  # Always False
+        for deployment in cluster.list_tool_deployments(user):
+            _, version = deployment.metadata.labels["chart"].rsplit("-", 1)
+            filter = filter | (
+                Q(chart_name=deployment.metadata.labels["app"]) &
+                Q(version=version)
+            )
+
+        tools = Tool.objects.filter(filter)
```

Comment:
I think this is the tricky bit as we get the supported tools from the DB but them map them to the deployments.
When there isn't corresponding record for the version the user has we don't return it.

If I try to remove the `version=` from the filter and from `cluster.list_tool_deployments()` things gets weird.

I don't know what's the best solution to keep showing unsupported/old versions of a tool in the list (and allowing user to restart, upgrade, etc...). Maybe you have some ideas regarding this, probably worth chat as I can imagine my comments may sound a bit random, sorry.

Reply:
One possibility would be have deployments in the list as they come from kubernetes (potentially augment with data coming from DB if the tool is of the same version) and use the DB mainly for deployment, e.g. as "lists of tools/versions" we support).

That would mean the `ToolDeployment` doesn't [incapsulate an instance of a `Tool`](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/732/files#diff-1b16cf3d380938c07fe469d8d7058f70R78) or possibly it's optional or something (but will have to be there when deploying).

Reply:
Yeah, I think getting the metadata from Kubernetes is the right approach. I'll have a try at making Kubernetes the source of truth instead of the db.

Reply:
Yeah, that may work. Sorry for find this and all the comments 😢 - feel free to Slack me if you need to discuss anything.

Reply:
No need to apologise, this is a good code review! The Tool version stuff isn't meant to be complete, but I realise now that it will actually break for some current users, so I'll disable for this PR and let you sort it out in yours :trollface:

Reply:
Yeah, as said I like the high level idea and the way this is structured but unfortunately there will be always users with older versions and hiding their tools is not gonna work ☺️

Everything will collapse the moment I touch anything 😱 😋